### ISSUE TITLE
Rich text editor

### ISSUE THREAD

**Issue by jesstelford** [keystonejs/keystone#616]
This is a consolidation of issues #10/#316/#377/#593 and new work by @mitchellhamilton.

Modern CMSs have _block_-based editing experiences which provide structure to input data beyond just HTML or markdown, but which also allow for freedom in the way those blocks are pieced together.

Keystone 5 needs a block-based editing experience.

> **Note: For motivations, see [this comment below](https://github.com/keystonejs/keystone-5/issues/616#issuecomment-497526661)**

There are internal needs that will be met by this style of editor over, say, a markdown or html field type.

@mitchellhamilton has been making great progress on such an editor and will provide more details below 👇 

See [comments below](https://github.com/keystonejs/keystone-5/issues/616#issuecomment-458823638) for:

- How to handle KS5 native types as Blocks (similar to `bold`/`link`/etc)
- How to model `Relationship`s
- How to store the data produced by the editor
- How to parse & render that data

## TODO <sup>in no particular order</sup>

- [x] **GraphQL Types**
  - [x] Generate internal join tables based on the block types added
  - [x] Generate input types for mutations
  - [x] Blocks generate GraphQL `input` types to be added as child fields on the `Content` field type
    For example;
    ```graphql
    {
      document
      relationshipUsersMention: relationshipUser(where: { type: 'mention' }) {
        username
      }
      relationshipUsersEmbed: relationshipUser(where: { type: 'embed' }) {
        username
        avatar
        bio
      }
    }
    ```
- [x] **Mutations**
  - [x] Client: Serialise Slate's document data before sending to the server (see https://github.com/keystonejs/keystone-5/issues/616#issuecomment-458831325)
  - [x] Server: Parse the serialised data on the server to perform relationship creates/connects
  - [x] Server: Inject resolved create/connect `id`s into serialised document before saving to DB
- [ ] **Queries**
  - [x] Client: Specify shape of data to return for complex blocks with a graphQL query
  - [x] Server: Parse serialised document to extract related IDs
  - [x] Server: Perform queries for the given IDs
  - [x] Client: Allow client to mutate data before injection
    For example;
    ```javascript
    query(gql`
      {
        document
        relationshipUsersMention: relationshipUser(where: { type: 'mention' }) {
          username
        }
        relationshipUsersEmbed: relationshipUser(where: { type: 'embed' }) {
          username
          avatar
          bio
        }
      }
    `).then(data => {
      return {
        ...data,
        relationshipUsers: data.relationshipUsersMention.concat(data.relationshipUsersEmbed),
      }
    });
    ```
    Which gets it back into the format the deserialisation lib is expecting.
  - [x] Client: Inject resolved data into serialised document
  - [ ] Switch to `graphql-type-json` for the slate document (#685)
- [x] **`CloudinaryImage` block type**
  - [x] Move `fields/types/Content/views/editor/blocks/image-container.js` to `fields/types/CloudinaryImage/views/block.js`
  - [x] Add support for rendering `{ data: { publicUrl: '...' } }`
- [ ] **`Relationship` block type**
  - [ ] ...
- [ ] **Tests**
- [x] **Documentation**

**Comment by Reinmar** [comment]
Hi 👋CKEditor 5 project lead here.

I've been recently researching how structured content editing could be implemented in CKEditor 5 for a talk I was preparing for [NeosCon](https://neoscon.io) and I found this ticket.

The PoC that I built on top of CKEditor 5 looks really promising and it solves the biggest UX problems that Neos CMS has with their current approach (multiple small editor instances for the editable parts of the content).

I made a demo of that prototype (together with a quick intro to CKEditor 5 itself) on NeosCon: https://twitter.com/reinmarpl/status/1133735637578854401.

The demo was tailored for Neos's reality. However, in order to succeed such a project needs to stay generic so it can be used by many CMSes.

Therefore, I'm curious about your requirements regarding structured content editing and potential issues if you already tried some solutions (not sure from the ticket what's the progress).

Thanks in advance for all the feedback! 🙂

**Comment by jesstelford** [comment]
Hey @Reinmar, thanks for sharing your work here!

The talk was very interesting; particularly the proof of concept you showed migrating from Neos existing editor made up of many different parts into a single unified block based editor - this is the exact use case that we're solving for in Keystone 5 too :)

We've been building on top of Slate.js to date, and will complete that work for this editor, but the great thing about Keystone 5 Field Types is it's up to the application developer to pick which one works for them, and I'd love to be able to offer a CKEditor Field Type (we already have a TinyMCE field, and plan to add a Markdown one also).

In building out this experience, the trickiest part I've encountered is serialisation of the editors data when saving to the DB. Probably best illustrated with a psuedo-example:

Imagine the editor gives us some data structure like so:

```javascript
{
  blocks: [
    {
      type: 'paragraph',
      blocks: [
        {
          type: 'image',
          data: { url: 'https://placekitten.com/g/2048/600' }
        }
      ]
    }
  ]
}
```

We can store that in the database easily enough, and enable querying it from our GraphQL API with something like:

```graphql
query {
  Post(id: "...") {
    content {
      document
    }
  }
}
```

Where `content.document` is the JSON structure from above which can be fed back into the editor for display.

Now imagine a display of the content for a mobile device; that image is much too large for a small screen. We could parse the document when we save it to the database and make multiple versions; one per device size, but as we add more variants the data gets very big very quickly, and we have to ensure data consistency across each variant, etc.

Instead, what we can do is allow the client to craft a graphQL query which modifies the data at read-time:

```graphql
query {
  Post(id: "...") {
    content {
      document
      images(transformation: { width: "700", height: "200" })
    }
  }
}
```

The data is now returned in a normalised format such as:

```javascript
{
  document: {
    blocks: [
      {
        type: 'paragraph',
        blocks: [
          {
            type: 'image',
            // The data is normalised
            data: { _joinId: 'abc213' }
          }
        ]
      }
    ]
  },
  images: [
    // NOTE: The url matches the requested image size
    { id: 'abc123', url: 'https://placekitten.com/g/700/200' }
  ]
}
```

Then after requesting the data, we can re-constitute it back into the format the editor expects ready for display by denormalising based on the `_joinId` & `type` fields for each block:

```javascript
{
  blocks: [
    {
      type: 'paragraph',
      blocks: [
        {
          type: 'image',
          data: { url: 'https://placekitten.com/g/700/200' }
        }
      ]
    }
  ]
}
```

Other possibilities this enables is, say, a query to get a gallery of all images in a post:

```graphql
query {
  Post(id: "...") {
    content {
      # NOTE: we don't query for `document`, only the images, and only at thumbnail size
      images(transformation: { width: "90", height: "90" })
    }
  }
}
```

Or for more complex examples:

```graphql
# Get all the users mentioned
query {
  Post(id: "...") {
    content {
      users {
        id
        username
        avatar
      }
    }
  }
}
```

```graphql
# Do a lookup to find all Posts a User 'def345' is mentioned in
# And extract out only the headline block (if set)
query {
  allPosts(where: {
    content: {
      users_contains: {
        id: "def345"
      }
    }
  }) {
    id
    content {
      headline {
        text
      }
    }
  }
}
```

I wonder if this is something you've thought about for the CKEditor? I know it's far beyond the editor's responsibility, but certain decisions when creating the editor's data structure can impact how easy it is to build these kinds of features.

**Comment by gautamsi** [comment]
I found https://editorjs.io/ with 9.6k stars.
@jesstelford  have you considered this before, if any drawbacks, let me know. I am planning to create this as field type for contrib (if not core)

**Comment by MadeByMike** [comment]
@jesstelford can we get a status update on the blocks\content editor? 

This issue has become really long and hard to track. It looks like the outstanding issues are related to blocks only? I'd like to review this issue and close it.

If there is any outstanding work on the content field type I suggest we consolidate this into a single issue and close these as duplicates: #650, #685, #728. There are also some minor UI inconsistencies that need attention and it would be good to see this captured if we're making a new issue.

If you don't have any time for this soon book in a short discussion with me and I'll see what I can do. I just need your context and history.

**Comment by ralexrdz** [comment]
Would love to be able to add images and store them as files, instead of data:images in the database. Something like LocalImage block

**Comment by MadeByMike** [comment]
Almost every item here has been complete. This grab-bag issue is probably no longer service purpose and future issues can be raised for additional WYSIWYG features
